### DISCUSSION TITLE
Does location-based AR.js work currently?

### DISCUSSION THREAD

**Discussion by ediejm** [AR-js-org/AR.js#255] [Q&A]
Hello - I've posted about this in a few places, so sorry if you've seen me before. 

I have been having significant trouble for a long time getting a location-based project to work. 

I have followed the docs exactly, but my objects continue to stick to the camera/device, rather than the GPS location defined. You can't walk around an object or towards/away from it. 

I have even tried using the exact code from the docs without a single change, or some sample code that Nicolo sent me, and they produced the same result. I have tried on Android and iOS devices, Safari and Chrome. 

I am aware that @nickw1 has geolocated AR.js working effectively in Hikar, so it makes me think that it must be possible to get it working. But if I have used the example code exactly and had no luck, I'm not sure what to try next. Are there alternatives I could try? This is the only markerless webAR tool I am aware of.

**Comment by nickw1** [comment]
@EchoDeltaMoon it should work, yes. Just a thought, can you test out

https://hikar.org/webapp/

on your device?

This is the Hikar app, which definitely works. If it does not work for you, it may be an issue with your device.

Also, if you are near any hills (within a few kilometres - they need not be mountains, just visible hills):

https://hikar.org/peakfinder/

**Reply by nickw1** [reply]
@EchoDeltaMoon sorry, I forgot to note: Hikar is only available in Europe and Turkey, so you need to be in these regions for it to work. (This is not due to any fundamental limitation, just due to the capacity constraints of my server)

**Reply by ediejm** [reply]
@nickw1 thank you so much for your reply. Yes, Hikar **does** work for me! Which is why I am so baffled that the other examples don't. I even tried copying the exact <a-camera> info from the Hikar repository and ensured I'm referencing the same versions of AR.js and A-frame etc. I think I am going mad...

**Reply by nickw1** [reply]
@EchoDeltaMoon ok, really strange. Are you using a simulated location or the real GPS location?
What happens if you scale down the object , I'm wondering whether it's a large, but distant, object, so that if you move a little way it will not noticeably change size?

Try posting the exact example code you're using, there are several basic examples available so it would be good to see which one you are using. Then I can see if the code works for me. Thanks.

**Reply by ediejm** [reply]
Thanks Nick. Here is the very basic sample with the box that I have tested. I am doing it with real co-ordinates for my location (hence removing them from the code below so I don't dox myself!) I have a screen record I can share if that helps (although I don't want to take up all your time troubleshooting for me!!) - showing how the box follows me as I move rather than staying in one spot, which I presume it is meant to.

```
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Test</title>
  <script src="https://aframe.io/releases/1.0.4/aframe.min.js"></script>
  <script src="https://unpkg.com/aframe-look-at-component@0.8.0/dist/aframe-look-at-component.min.js"></script>
  <script src="https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js"></script>
  <script src="https://cdn.jsdelivr.net/gh/donmccurdy/aframe-extras@v6.1.1/dist/aframe-extras.min.js"></script>
</head>

  <body style="margin: 0; overflow: hidden;">
    <a-scene
      vr-mode-ui="enabled: false"
      embedded
      arjs="sourceType: webcam; debugUIEnabled: false;"
    >

    <a-box
    gps-entity-place="latitude: <lat>; longitude: <long>;"
    scale="10 10 10"
    ></a-entity>

      <a-camera gps-camera rotation-reader> </a-camera>
    </a-scene>
  </body>
</html>
```

**Reply by nickw1** [reply]
Hi @EchoDeltaMoon ok thanks. One thing I have norticed is that you open `<a-box>` and close `<a-entity>`, I presume this is a mistake, this could cause it not to work as expected.

Is this an error in the example? If so,  apologies, let me know which example this was based on.

**Reply by ediejm** [reply]
Oops yes thank you for that spot - it's my fault, because I had an object but I thought I'd change to a box to check it wasn't my model causing the problem. The rest is from https://ar-js-org.github.io/AR.js-Docs/ but with a box instead of text. But I am about to give it a try with the text.

**Reply by ediejm** [reply]
Yes same issue with the code at https://ar-js-org.github.io/AR.js-Docs/ sadly :(

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <title>GeoAR.js demo</title>
    <script src="https://aframe.io/releases/1.0.4/aframe.min.js"></script>
    <script src="https://unpkg.com/aframe-look-at-component@0.8.0/dist/aframe-look-at-component.min.js"></script>
    <script src="https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js"></script>
  </head>

  <body style="margin: 0; overflow: hidden;">
    <a-scene
      vr-mode-ui="enabled: false"
      embedded
      arjs="sourceType: webcam; debugUIEnabled: false;"
    >
      <a-text
        value="This content will always face you."
        look-at="[gps-camera]"
        scale="120 120 120"
        gps-entity-place="latitude: <add-your-latitude>; longitude: <add-your-longitude>;"
      ></a-text>
      <a-camera gps-camera rotation-reader> </a-camera>
    </a-scene>
  </body>
</html>
```

**Reply by ediejm** [reply]
![200](https://user-images.githubusercontent.com/72398292/116446851-477a1d00-a84f-11eb-99e4-c3784bdd5109.gif)

I tried to get a screen record that shows it but you can probably barely see....but it sticks to the camera as I walk around

**Reply by nickw1** [reply]
OK try swapping the camera and the text round, also change the text to a box (text will still work, but box is easier to test because it's 3D).

**Reply by nickw1** [reply]
... sorry, to follow up with that. Swapping the camera makes no difference. I am finding the example works on my phone in the sense that if I change the lat and lon in the HTML to be further away, the box gets smaller. Soon (before it gets dark - UK timezone) I will go outside and see if it works. If it does I will post my code with my lat/lon anomymised.

**Reply by ediejm** [reply]
Thank you, I really appreciate you taking the time to help me out!

**Reply by nickw1** [reply]
@EchoDeltaMoon ok this example definitely works on my phone (Android 10, latest Chrome, Pixel 3)
where `x` is my current longitude and `y` my current latitude. The boxes definitely get bigger as you approach them, smaller as you move away and appear in the correct geographical location.

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <title>GeoAR.js demo</title>
    <script src="https://aframe.io/releases/1.0.4/aframe.min.js"></script>
    <script src="https://unpkg.com/aframe-look-at-component@0.8.0/dist/aframe-look-at-component.min.js"></script>
    <script src="https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js"></script>
  </head>

  <body style="margin: 0; overflow: hidden;">
    <a-scene
      vr-mode-ui="enabled: false"
      embedded
      arjs="sourceType: webcam; debugUIEnabled: false;"
    >
      <a-box 
        look-at="[gps-camera]"
		material="color: yellow"
        scale="10 10 10"
        gps-entity-place="longitude: y-0.0005; latitude: x+0.0006;"
      ></a-box>
      <a-box 
        look-at="[gps-camera]"
		material="color: red"
        scale="10 10 10"
        gps-entity-place="longitude: y-0.0001; latitude: x+0.0004;"
      ></a-box>
      <a-camera gps-camera rotation-reader> </a-camera>
    </a-scene>
  </body>
</html>
```

**Reply by nickw1** [reply]
@EchoDeltaMoon sorry `x` and `y` are the wrong way round. `x` is longitude and `y` is latitude.

**Reply by ediejm** [reply]
Tysm! Have tried it and it seems to sort've work and sort've not at the same time but I will test it properly in the daytime tomorrow. Tysm again for your help and patience.

PS does the number of decimal places in the lat/long matter? Is it 'the more the better' or does it break it if you have too many?

**Reply by nickw1** [reply]
It doesn't matter too much but if you use offsets of around 0.0005 from your current location (as in my example) then the objects are a sensible distance away, perhaps 20-50 metres, the sort of distance that works well in testing.
